mesonbuild/build.py

```diff
@@ -474,7 +474,21 @@ def process_compilers_late(self):
                     self.compilers[lang] = compilers[lang]
                     break
 
-    def process_compilers(self):
+    def cyclic_dep_breaker_for_depsources(self, external_dependencies):
+        result = []
+        # We need to have external dependencies here because they may contain
+        # sources but we can only process kwargs _after_ this function is
+        # done because D requires code from this. The correct solution would
+        # be to fix D compiler to not need this.
```

Comment:
I don't understand this comment, what code does D require and where?

Reply:
The proper fix would be moving `process_kwargs` on line 368 above `process_compilers` (or something of the sort, I forget offhand). Because the order is the current one, compilers need to be defined before processing kwargs because it converts one keyword argument into a compiler flag immediately rather than storing the intention. This causes D test number 9 to fail when it's building `d_import_dirs`. The correct thing here would probably be to convert it to an include_directories object.

@ximion: since you know D best, why is that using `d_import_dirs` rather than regular include_directories? Is there a conceptual difference between the two?

Reply:
@jpakkane D's import_dirs is a D specific feature not comparable at all with `include_directories`. Include directories reference directories to include headers, while a string import dir contains auxiliary non-D textfiles or other data that is compiled into the binary at compile time or interpreted (this e.g. allows D to parse other languages at compile time or do other fancy stuff). All D compilers use the `-J` flag for those directories at time.
See https://github.com/mesonbuild/meson/blob/master/test%20cases/d/9%20features/app.d#L14 for a dumb code example of how one would use the feature.

Reply:
Regardless, people should not need to build paths to source directories manually. That is pointless busywork that we should do on their behalf. Either we should expand `foobar` ourselves or (also) make it accept directories specified with `include_directories`.

Reply:
While I agree in concept, there is no way we know which directories D sources should be included from and from which directories we want to include string imports. Those can have annoying filename clashes and I see huge potential to make the life of D programmers miserable by constantly including every regular include and source dir as string import dir as well. Maybe adding a flag to the include_dir objects makes sense, but I am unsure about that...
In most D projects, string imports are used sparingly, it's not something you will find in most D sources (some projects use it heavily though, like the Diet template module of the Vibe.d web framework).

Reply:
There would still be different ways to input each. Source searches would use `include_directories` keyword and string imports would use `d_import_dirs`, it's just a question of saving people the need to join directories manually. (`include_directories` function is basically just a glorified version of "remember this directory location end expand properly once it is used").

Reply:
Ah, that sounds good - I must admit, I haven't looked at this patch in general, and basically only answered the question that was asked.
Sorry for the confusion :-)